### PR TITLE
task-cgroup --> dev (b2b)

### PR DESCRIPTION
**Pull request by vsiddharth**
### Summary

Enable task level resource (CPU, Memory) constraints

### Implementation details

Hierarchical cgroups are the cornerstone of this implementation.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes:  YES

### Description for the changelog
Feature - Enable task level CPU and memory constraints

### Licensing

This contribution is under the terms of the Apache 2.0 License:  YES

### CODE REVIEW

File: agent/acs/model/api/api-2.json [hunk 1]
```diff
@@ -193,16 +194,6 @@
         "attachmentArn":{"shape":"String"},
         "ec2Id":{"shape":"String"},
         "ipv4Addresses":{"shape":"IPv4AddressList"},
-        "ipv6Addresses":{"shape":"IPv6AddressList"}
```
Comment (richardpen) [inline agent/acs/model/api/api-2.json]: Why is this removed?

File: agent/statemanager/state_manager.go [hunk 2]
```diff
@@ -52,7 +52,9 @@ const (
 	// 7)
 	//   a) Add 'MetadataUpdated' field to 'api.Container'
 	//   b) Add 'DomainNameServers' and 'DomainNameSearchList' in `api.ENI`
-	ECSDataVersion = 7
+	// 8)
+	//   a) Added task cgroup related fields to 'api.Task'
```
Comment (aaithal) [inline agent/statemanager/state_manager.go]: Can you please list these fields as well?

Comment (vsiddharth) [reply agent/statemanager/state_manager.go]: I've resolved this later on where I've explicitly mentioned all the new fields.

File: agent/acs/model/api/api-2.json (hunk 1 above)
Comment (richardpen) [reply agent/acs/model/api/api-2.json]: Ignore it, it's hidden.

**Comment by vsiddharth** [comment]
@aaithal I've updated the `CHANGELOG`. Please have another look when you have the time!

**Comment by aaithal** [comment]
@vsiddharth can you please resolve the merge conflict? Everything looks good to me.

**Comment by vsiddharth** [comment]
@aaithal This has been manually tested several times in the past and I've also tested this feature end-to-end with the latest changes.

**Comment by vsiddharth** [comment]
This change needs to be coupled with this `amazon-ecs-init` [PR](https://github.com/aws/amazon-ecs-init/pull/120)

**Comment by vsiddharth** [comment]
@sharanyad I had resolved all conflicts earlier in the day but more commits have been pushed to `dev`. I'll resolve them yet again!

**Comment by vsiddharth** [comment]
The current commit view is confusing with merge commits. 
Please have a look [here](https://github.com/aws/amazon-ecs-agent/compare/dev...vsiddharth:dev-task-cgroup-b2b?expand=1) for additional clarity.
